File: Examples/test-suite/perl5/director_binary_string_runme.pl
use strict;
use warnings;
use Test::More tests => 4;
BEGIN { use_ok 'director_binary_string' }
require_ok 'director_binary_string';

{
  package DirectorBinaryStringCallback;
  use base 'director_binary_string::Callback';
  sub run {
      my $ret = 0;
      my ($self, $dataBufferAA, $dataBufferBB) = @_;
      if(defined $dataBufferAA) {
        $ret += ord($_) * 2 for(split('', $dataBufferAA));
      }
      if(defined $dataBufferBB) {
        $ret += ord($_) * 3 for(split('', $dataBufferBB));
      }
      return $ret;
  }
  sub writeData {
      my $ret = 0;
      my ($self, $dataBufferAA) = @_;
      if(defined $dataBufferAA) {
        $ret += ord($_) * 2 for(split('', $dataBufferAA));
      }
      return $ret;
  }
}

{
  print "Start\n";
  my $caller = director_binary_string::Caller->new();
  my $callback = DirectorBinaryStringCallback->new();
  $caller->setCallback($callback);
  my $sum = $caller->call();
  my $sumData = $caller->callWriteData();
  $caller->delCallback();

  is($sum, 9*2*8 + 13*3*5, 'Unexpected sum: ' . $sum);
  is($sumData, 9*2*8, 'Unexpected sumData: ' . $sumData);

# FIXME accessing abstract class or protected constructor
# It does make sense as writeData() is abstract
# director_binary_string::Callback->new()->run(undef, undef);
  $callback = DirectorBinaryStringCallback->new();
  $caller->setCallback($callback);
  $caller->call_null();
  $caller->delCallback();
}

File: Examples/test-suite/octave/director_binary_string_runme.m
director_binary_string

function self=DirectorBinaryStringCallback()
  global director_binary_string;
  self=subclass(director_binary_string.Callback());
  self.run=@DirectorBinaryStringCallback_run;
  self.writeData=@DirectorBinaryStringCallback_writeData;
end
function ret=DirectorBinaryStringCallback_run(self, dataBufferAA, dataBufferBB)
  ret = 0;
  for i = 1:length(dataBufferAA)
    ret = ret + double(dataBufferAA(i)) * 2;
  end
  for i = 1:length(dataBufferBB)
    ret = ret + double(dataBufferBB(i)) * 3;
  end
end
function ret=DirectorBinaryStringCallback_writeData(self, dataBufferAA)
  ret = 0;
  for i = 1:length(dataBufferAA)
    ret = ret + double(dataBufferAA(i)) * 2;
  end
end

caller = director_binary_string.Caller();
callback = DirectorBinaryStringCallback();
caller.setCallback(callback);
sum = caller.call();
sumData = caller.callWriteData();
caller.delCallback();

if (sum != 9*2*8 + 13*3*5)
    error(sum);
end
if (sumData != 9*2*8)
    error(sumData);
end

% FIXME how do we create null string?
% director_binary_string.Callback().run(null, null);
director_binary_string.Callback().run('', '');
callback = DirectorBinaryStringCallback();
caller.setCallback(callback);
caller.call_null();
caller.delCallback();
